gnomad_qc/v4/annotations/generate_freq.py

```diff
@@ -0,0 +1,441 @@
+    SYNDIP,
+)
+
+logging.basicConfig(
+    level=logging.INFO,
+    format="%(asctime)s: %(message)s",
+    datefmt="%m/%d/%Y %I:%M:%S %p",
+)
+logger = logging.getLogger("gnomAD_frequency")
+logger.setLevel(logging.INFO)
+
+
+def annotate_non_ref_het(vds: hl.vds.VariantDataset) -> hl.vds.VariantDataset:
+    """
+    Annotate non-ref heterozygous calls to prevent incorrect call adjustment post-split.
+
+    :param vds: Hail VDS to annotate non-ref het onto variant data.
+    :return: Hail VDS with _non_ref_het annotation.
+    """
+    vds.variant_data = vds.variant_data.annotate_entries(
+        _het_non_ref=vds.variant_data.LGT.is_het_non_ref()
+    )
+    return vds
+
+
+def annotate_high_ab_hets_by_group_membership(
+    mt: hl.MatrixTable,
+    ab_cutoff: hl.float = 0.9,
+) -> hl.MatrixTable:
+    """
+    Annotate high AB hets by group membership.
+
+    :param mt: MatrixTable to annotate high AB hets onto.
+    :param ab_cutoff: Allele balance cutoff for hom alt depletion fix. Defaults to 0.9
+    :return: _description_
+    """
+    logger.info("Annotating number of high AB het sites in each freq group...")
+    mt = mt.annotate_rows(
+        high_ab_hets_by_group_membership=hl.agg.array_agg(
+            lambda i: hl.agg.filter(
+                mt.group_membership[i]
+                & needs_high_ab_het_fix_expr(
+                    mt,
+                    ab_cutoff,
+                ),
+                hl.agg.count(),
+            ),
+            hl.range(hl.len(mt.group_membership)),
+        )
+    )
+    return mt
+
+
+def needs_high_ab_het_fix_expr(
+    mt: hl.MatrixTable,
+    ab_cutoff: hl.float = 0.9,
+) -> hl.MatrixTable:
+    """
+    Annotate the MatrixTable rows with the count of high AB het genotypes per frequency sample grouping.
+
+    This arrary allows the AC to be corrected with the homalt depletion fix.
+
+    :param mt: Hail MatrixTable to annotate.
+    :param ab_cutoff: Alelle balance threshold at which the GT changes to homalt. Default is 0.9.
+    :return mt:
+    """
+    return (
+        (mt.AD[1] / mt.DP > ab_cutoff)
+        & mt.adj
+        & mt.GT.is_het_ref()
+        & mt.meta.project_meta.fixed_homalt_model
+        & ~mt._het_non_ref  # Skip adjusting genotypes if sample originally had a het nonref genotype
+    )
+
+
+def subtract_high_ab_hets_from_ac(
+    mt: hl.MatrixTable, af_threshold: float = 0.01
+) -> hl.Table:
+    """
+    Correct frequencies at sites with an AF greater than the af_threshold.
+
+    :param ht: Hail Table containing freq and high_ab_het annotations.
+    :param af_threshold: AF threshold at which to correct frequency. Default is 0.01.
+    :return: Hail Table with adjusted frequencies.
+    """
+    mt = mt.annotate_rows(
+        ab_adjusted_freq=hl.if_else(
+            mt.freq[0].AF < af_threshold,
+            mt.freq,
+            hl.map(
+                lambda f, g: hl.struct(
+                    AC=hl.int32(f.AC - g),
```

Comment:
Maybe I'm totally thinking about this wrong, but don't we want to add? We are fixing a hom alt depletion, so we expect a higher AC

Reply:
Nope, you are correct